**Jayasankar J N**  <br />
S5 CSE <br /> Roll No: 39 <br /> TVE19CS039

Loading the data and splitting it into train and test sets

In [1]:
import tensorflow as tf

mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# x_train: (60000, 28, 28)
# y_train: (10000, 28, 28)

x_train, x_test = x_train / 255, x_test / 255

11501568/11490434 [==============================] - 0s 0us/step


# Model 1
Here the layers are: 1 input layer having 784 neurons, 1 hidden layer having 25 units and 1 output layers having 10 units (10 classes). The count of total parameters is 19885 which is much less than the training size of 60000.
Hidden layer uses 'ReLU' activation function and output layers uses softmax.
Model is trained for 5 epochs.

In [9]:
model1 = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(25, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy()

model1.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])
model1.summary()

print("Training...")
model1.fit(x_train, y_train, epochs=5)

print("Testing...")
model1.evaluate(x_test, y_test)

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_7 (Flatten)         (None, 784)               0         
                                                                 
 dense_14 (Dense)            (None, 25)                19625     
                                                                 
 dense_15 (Dense)            (None, 10)                260       
                                                                 
Total params: 19,885
Trainable params: 19,885
Non-trainable params: 0
_________________________________________________________________
Training...
Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3805 - accuracy: 0.8934
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2306 - accuracy: 0.9349
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1920 - accuracy: 0.9445

[0.1507243663072586, 0.95169997215271]

Model 1:

*   Number of epochs = 5 
*   Training Accuracy = 0.9575
*   Test Accuray = 0.9517




# Model 2

Layers are: 1 input layer having 784 neurons, 2 hidden layers, one having 64 units and other having 32 units and 1 output layers having 10 units (10 classes).The count of total parameters is 52650 which is really close to the training size of 60000, so there is high chance that this model will suffer from overfitting.
Hidden layers uses 'ReLU' activation function and output layers uses softmax.
Model is trained for 20 epochs.

In [14]:
model2 = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(64, activation='relu'),
  tf.keras.layers.Dense(32, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy()

model2.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])
model2.summary()

print("Training...")
model1.fit(x_train, y_train, epochs=20)

print("Testing...")
model1.evaluate(x_test, y_test)

Model: "sequential_12"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_12 (Flatten)        (None, 784)               0         
                                                                 
 dense_28 (Dense)            (None, 64)                50240     
                                                                 
 dense_29 (Dense)            (None, 32)                2080      
                                                                 
 dense_30 (Dense)            (None, 10)                330       
                                                                 
Total params: 52,650
Trainable params: 52,650
Non-trainable params: 0
_________________________________________________________________
Training...
Epoch 1/20
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0362 - accuracy: 0.9883
Epoch 2/20
1875/1875 [==============================] - 4s 2ms/step

[0.2084774374961853, 0.9606999754905701]

Model 2:

*   Number of epochs = 20
*   Training Accuracy = 0.9927
*   Test Accuray = 0.9607
